src/billing_handler.rs

```diff
@@ -1,34 +1,63 @@
 use crate::model::AppState;
 
-use actix_web::web::Data;
-use actix_web::HttpResponse;
+use actix_web::web::{Data, Json};
+use actix_web::{get, post, HttpResponse, Responder};
 use anyhow::Context;
 use k256::elliptic_curve::generic_array::sequence::Lengthen;
 use serde_json::json;
 use tiny_keccak::{Hasher, Keccak};
 
-pub async fn billing_data(appstate: Data<AppState>) -> HttpResponse {
-    let mut execution_costs_gaurd = appstate.execution_costs.lock().unwrap();
+#[derive(Debug, serde::Deserialize)]
+pub struct SigningData {
+    nonce: String,
+    txhashes: Vec<String>,
+}
 
-    if execution_costs_gaurd.is_empty() {
-        return HttpResponse::BadRequest().body("No billing data available");
+#[get("/bill")]
+pub async fn get_bill(appstate: Data<AppState>) -> impl Responder {
+    let bill = appstate.execution_costs.lock().unwrap().clone();
+    if bill.is_empty() {
+        return HttpResponse::NoContent().body("No bill data available");
     }
 
-    let mut billing_data: Vec<u8> = Vec::new();
+    HttpResponse::Ok().json(json!({
+        "bill": bill,
+    }))
+}
+
+#[post("/sign")]
+pub async fn sign_data(appstate: Data<AppState>, data: Json<SigningData>) -> impl Responder {
+    let signing_data = data.into_inner();
+    if signing_data.nonce.is_empty() {
+        return HttpResponse::BadRequest().body("Nonce must not be empty");
+    }
 
-    for (tx_hash, cost) in execution_costs_gaurd.iter() {
-        let mut bytes32_tx_hash = [0u8; 32];
-        if let Err(err) = hex::decode_to_slice(&tx_hash[2..], &mut bytes32_tx_hash) {
-            return HttpResponse::InternalServerError()
-                .body(format!("Error decoding transaction hash: {:?}", err));
+    if signing_data.txhashes.is_empty() {
+        return HttpResponse::BadRequest().body("List of tx hashes must not be empty");
+    }
+
+    let mut signed_data: Vec<u8> = Vec::new();
+    for txhash in signing_data.txhashes {
+        if let Some(cost) = appstate.execution_costs.lock().unwrap().remove(&txhash) {
+            let mut bytes32_txhash = [0u8; 32];
+            if let Err(err) = hex::decode_to_slice(&txhash[2..], &mut bytes32_txhash) {
+                return HttpResponse::InternalServerError()
+                    .body(format!("Error decoding transaction hash: {:?}", err));
+            }
+
+            signed_data.append(&mut bytes32_txhash.to_vec());
+            signed_data.append(&mut cost.to_be_bytes().to_vec());
+        } else {
+            return HttpResponse::BadRequest().body(format!(
```

Comment:
Do not exit aggressively. Losing cost info for valid tx hashes. Either put rollback for execution costs of valid tx hashes or return sign for valid tx hashes only and missing tx hashes in separate field.

Reply:
I think it is not bad idea to call a rollback method whenever error is occurring in this method.

Reply:
Yes that's a good point. @roshanr95 What is the protocol in case some of the tx hashes in the list provided by the user are not present in the execution costs object? Throw out an error and call a rollback method to retrieve the removed tx hashes OR prepare the signature and billing data from the remaining valid tx hashes and return the missing tx hashes back in a separate field?

Reply:
Having a rollback method is a safe option. The only trade off here will be the extended lock we'll have to take on the execution costs object to add back the removed data, which can block other threads that are trying to update the bill after each serverless app call.
Even if we don't use a rollback method in the above case and simply return the missing tx hashes, we'll still need to add one for the case where signing the billing data results in an error. Also it can happen that all the tx hashes provided by the user are invalid so that's also there.

Reply:
Go with rollback then. It's not a significant trade-off, given correctness on the other hand.

Reply:
can't we just return 0, why do we care if the hash exists or not :man_shrugging:

Reply:
regardless, no point in erroring out here, either assume 0 cost or just exclude the tx hash from the final thing

Reply:
Will exclude the tx hash from the final data.